bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/handlers/TraversePageHandler.java

```diff
@@ -28,30 +28,84 @@
  * @since 3.5
  */
 public class TraversePageHandler extends WidgetMethodHandler {
-
 	/**
 	 * The parameters for traverse(int).
 	 */
 	private static final Class<?>[] METHOD_PARAMETERS = { int.class };
-
 	@Override
 	public final Object execute(final ExecutionEvent event) {
 		Control focusControl = Display.getCurrent().getFocusControl();
 		if (focusControl != null) {
-			int traversal = "next".equals(methodName) ? SWT.TRAVERSE_PAGE_NEXT : SWT.TRAVERSE_PAGE_PREVIOUS; //$NON-NLS-1$
+			boolean forward = "next".equals(methodName); //$NON-NLS-1$
+			int traversal = getTraversalDirection(forward);
 			Control control = focusControl;
+
 			do {
+				if (control instanceof CTabFolder folder && isFinalItemInCTabFolder(folder, forward)
+						&& !areHiddenItems(folder)) {
+					loopToSecondToFirstItemInCTabFolder(folder, forward);
+					traversal = getTraversalDirection(!forward); // we are in the second-to-last item in the given
+					// direction. Now, use the Traverse-event to move back by one
+				}
 				if (control.traverse(traversal))
 					return null;
 				if (control instanceof Shell)
 					return null;
 				control = control.getParent();
 			} while (control != null);
 		}
-
 		return null;
 	}
 
+	private boolean areHiddenItems(CTabFolder folder) {
+		CTabItem[] items = folder.getItems();
+		for (CTabItem i : items) {
+			if (!i.isShowing()) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+	private int getTraversalDirection(boolean direction) {
+		return direction ? SWT.TRAVERSE_PAGE_NEXT : SWT.TRAVERSE_PAGE_PREVIOUS;
+	}
+
+	/**
+	 * Sets the current selection to the second-to-last item in the given direction.
+	 *
+	 * @param folder  the CTabFolder which we want to inspect
+	 * @param forward whether we want to traverse forwards of backwards
+	 */
+	private void loopToSecondToFirstItemInCTabFolder(CTabFolder folder, boolean forward) {
+		if (forward) {
+			folder.showItem(folder.getItem(0));
+			folder.setSelection(1);
+		} else {
+			int itemCount = folder.getItemCount();
+			folder.setSelection(itemCount - 2);
+		}
+	}
+
+	/**https://github.com/jannisCode/eclipse.jdt.ui.git
```

Comment:
By convention, the first line of javadoc comments is empty (atleast that's what I have seen everywhere else)